GCS/probe0/probe0.go

```diff
@@ -0,0 +1,142 @@
+import (
+	"context"
+	"fmt"
+	"sync"
+	"io"
+	"os"
+	"time"
+	"cloud.google.com/go/storage"
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+	"github.com/google/cloudprober/targets/endpoint"
+	"github.com/googleinterns/step224-2020/blob/evan/hermesMain/cmd"
+)
+
+// Probe holds aggregate information about all probe runs, per-target.
+type Probe struct {
+	name    string
+	c       *ProbeConf
+	targets []string
+	opts    *options.Options
+
+	res map[string]*metrics.EventMetrics // Results by target
+	l   *logger.Logger
+}
+
+// Init initializes the probe with the given params.
+func (p *Probe) Init(name string, opts *options.Options) error {
+	c, ok := opts.ProbeConf.(*ProbeConf)
+	if !ok {
+		return fmt.Errorf("not a my probe config")
+	}
+	p.c = c
+	p.name = name
+	p.opts = opts
+	p.l = opts.Logger
+	p.res = make(map[string]*metrics.EventMetrics)
+
+	return nil
+}
+
+// Start starts and runs the probe indefinitely.
+func (p *Probe) Start(ctx context.Context, dataChan chan *metrics.EventMetrics) {
+	probeTicker := time.NewTicker(p.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			// On probe tick, write data to the channel and run probe.
+			for _, em := range p.res {
+				dataChan <- em
+			}
+			p.targets = endpoint.NamesFromEndpoints(p.opts.Targets.ListEndpoints())
+			p.initProbeMetrics()
+			probeCtx, cancelFunc := context.WithDeadline(ctx, time.Now().Add(p.opts.Timeout))
```

Comment:
Consider making cancelFunc a struct field for Probe, this would allow us to actually cancel the probe from outside